coreblocks/structs_common/instr_counter.py

```diff
@@ -0,0 +1,46 @@
+from amaranth import *
+from coreblocks.params.genparams import GenParams
+from coreblocks.params.layouts import CoreInstructionCounterLayouts
+from transactron.core import Method, TModule, def_method
+
+
+class CoreInstructionCounter(Elaboratable):
+    """
+    Counts instructions currently processed in core.
+    Used in exception handling, to wait for core flush to finish.
+
+    Attributes
+    ----------
+    increment : Method
+        Increments the counter. Should be called when new instruction leaves fetch stage.
+    decrement : Method
+        Decrements the counter, and returns if the counter will be equal to zero after that cycle (it was the
+        last instruction in core and no new instruction is fetched). Should be called when instruction is retired.
+    """
+
+    def __init__(self, gp: GenParams):
+        self.gp = gp
+
+        self.increment = Method()
+        self.decrement = Method(o=gp.get(CoreInstructionCounterLayouts).decrement)
+
+    def elaborate(self, platform) -> TModule:
+        m = TModule()
+
+        count = Signal(self.gp.rob_entries_bits + 1)
```

Comment:
What if someone will create a rob with 1 entry? Then we will have ~4 instructions before ROB, which will cause that the counter will overrun. (I know, this is a very weird edge case :D)

```suggestion
        # Make sure that we have enough bits in counter to count all instructions before ROB
        count = Signal(max(4,self.gp.rob_entries_bits + 1))
```

Reply:
done